src/main/java/net/minecraftforge/common/ForgeSpawnEggItem.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common;
+
+import com.google.common.collect.Iterables;
+import net.minecraft.core.BlockSource;
+import net.minecraft.core.Direction;
+import net.minecraft.core.dispenser.DefaultDispenseItemBehavior;
+import net.minecraft.core.dispenser.DispenseItemBehavior;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.world.entity.*;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.SpawnEggItem;
+import net.minecraft.world.level.block.DispenserBlock;
+import net.minecraft.world.level.gameevent.GameEvent;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+public class ForgeSpawnEggItem extends SpawnEggItem
+{
+    private static final List<ForgeSpawnEggItem> MOD_EGGS = new ArrayList<>();
+    private final Supplier<? extends EntityType<? extends Mob>> typeSupplier;
+
+    public ForgeSpawnEggItem(Supplier<? extends EntityType<? extends Mob>> type, int backgroundColor, int highlightColor, Properties props)
+    {
+        super((EntityType<? extends Mob>) null, backgroundColor, highlightColor, props);
+        this.typeSupplier = type;
+
+        MOD_EGGS.add(this);
+    }
+
+    @Override
+    public EntityType<?> getType(@Nullable CompoundTag tag)
+    {
+        EntityType<?> type = super.getType(tag);
+        return type != null ? type : typeSupplier.get();
+    }
+
+    @Nullable
+    protected DispenseItemBehavior createDispenseBehavior()
+    {
+        return DEFAULT_DISPENSE_BEHAVIOR;
+    }
+
+    public static Iterable<SpawnEggItem> eggs()
+    {
+        return Iterables.concat(SpawnEggItem.eggs(), MOD_EGGS);
```

Comment:
This is going to result in many duplicates, depending on if called before or after finalize. Better to call distinct.
Or, we could not use our own map, and just BY_ID as our map. Or, we could not merge things at all.
There would be two places where we would care about it.
IForgeEntity.getPickedResult and Mob.getPickedResult, both of which are fine with null, so you should change it to checking both maps.

Reply:
The picked results don't care about this specific method at all.
If I move the color registration to an event, then I could remove the concatanation and just return the `MOD_EGGS` list because the only use for that list outside of the `ForgeSpawnEgg` class is the color handler registration. In this case I also wouldn't need to clear the list.

Reply:
This method now only returns eggs registered by mods and is used by the color handler registration